## <u> Feature Preparataion </u>

This notebook covers feature engineering and feature selection for the Home Credit Default Risk model. While using all 100 available features is computationally feasible, deliberate dimensionality reduction improves model performance by removing noise and enhances interpretability by clarifying each feature's contribution to default risk. This follows the exploratory analysis from the previous notebook, applying those insights to construct a refined feature set optimized for both predictive accuracy and business utility.

### <u> 1. Setup</u>


In [1]:
# Import libraries
import pandas as pd, numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sys
import phik

# Set root directory for module imports
sys.path.append('..')

# Import modules
from modules.modules_eda import *


In [2]:
# Import dataset and column descriptions
home_credit = pd.read_csv('../data/processed/home_credit_cleaned.csv', index_col = 0)
columns_info = pd.read_csv('../data/processed/columns_info.csv')
print(f'Shape of the dataset: {home_credit.shape}')

Shape of the dataset: (307511, 100)


### <u> 2. Phi-K Correlation</u>

Given the different nature of the variables (binaries, categorical and continuous), standard Pearson correlation is inappropriate for feature selection as it assumes linearity and continuous distributions. As a solution to address this limitation, phi-k correlation is implemented as a substitute. Contrary to Pearson correlation, phi-k evaluates association strength across mixed variable types on a 0-1 scale by comparing the observed joint distribution against what would be expected under statistical independence. This allows meaningful comparison of feature relevance regardless of whether variables are binary, categorical, or continuous, allowing a first selection before feature engineering procedures.

In [3]:
# Calculate phik correlation matrix
#phik_corr = home_credit.phik_matrix()

# Save phik correlation matrix
# phik_corr.to_csv('../data/processed/phik_correlation_matrix.csv')

In [4]:
# Extract target correlations
phik_corr = pd.read_csv('../data/processed/phik_correlation_matrix.csv', index_col = 0)
target_corr = phik_corr['TARGET'].sort_values(ascending = False).drop('TARGET')
target_corr = target_corr.reset_index()
target_corr.index = target_corr.index + 1
target_corr.index.name = 'importance_rank'
target_corr.columns = ['feature', 'phik corr. TARGET']

target_corr.head(50)

feature  phik corr. TARGET
importance_rank                                                  
1                                 EXT_SOURCE_3           0.247680
2                                 EXT_SOURCE_1           0.217846
3                                 EXT_SOURCE_2           0.213965
4                               YEARS_EMPLOYED           0.103535
5                                  YEARS_BIRTH           0.102328
6                              OCCUPATION_TYPE           0.090029
7                            ORGANIZATION_TYPE           0.089164
8                             NAME_INCOME_TYPE           0.084831
9                       REG_CITY_NOT_WORK_CITY           0.079946
10                     YEARS_LAST_PHONE_CHANGE           0.073182
11                              FLAG_EMP_PHONE           0.072087
12                      REG_CITY_NOT_LIVE_CITY           0.069588
13                             FLAG_DOCUMENT_3           0.069525
14                              bur_cnt_active           0.068030
15                            YEARS_ID_PUBLISH           0.067829
16                             AMT_GOODS_PRICE           0.059093
17                                  AMT_CREDIT           0.056769
18                          NAME_FAMILY_STATUS           0.056043
19                                 OWN_CAR_AGE           0.055357
20                          YEARS_REGISTRATION           0.054582
21               AMT_REQ_CREDIT_BUREAU_flag_na           0.051462
22                           NAME_HOUSING_TYPE           0.051107
23                     LIVE_CITY_NOT_WORK_CITY           0.050956
24                          NAME_CONTRACT_TYPE           0.048399
25                             bur_has_history           0.048237
26                         NAME_EDUCATION_TYPE           0.046954
27                             FLAG_DOCUMENT_6           0.044791
28                             FLAG_WORK_PHONE           0.044678
29                  REGION_POPULATION_RELATIVE           0.042570
30                  prev_cnt_consumer_approved           0.038672
31                                  FLAG_PHONE           0.037258
32                 REGION_RATING_CLIENT_W_CITY           0.036699
33                        REGION_RATING_CLIENT           0.035450
34                     HOUR_APPR_PROCESS_START           0.034577
35                                FLAG_OWN_CAR           0.034179
36                                 CODE_GENDER           0.032921
37                                 AMT_ANNUITY           0.029861
38                  prev_cnt_revolving_refused           0.029505
39                            prev_has_history           0.028838
40                 prev_cnt_revolving_canceled           0.028548
41                       prev_cnt_cash_refused           0.028228
42                    DEF_60_CNT_SOCIAL_CIRCLE           0.018315
43                            FLAG_DOCUMENT_16           0.017927
44                            FLAG_DOCUMENT_13           0.017812
45                       prev_cnt_xna_canceled           0.017355
46                 prev_cnt_revolving_approved           0.017223
47                   AMT_REQ_CREDIT_BUREAU_MON           0.016467
48                             CNT_FAM_MEMBERS           0.016143
49                            FLAG_DOCUMENT_14           0.014416
50                      prev_cnt_cash_approved           0.013466

<u> **Comment:** </u>

As we can observe from the sorted correlation rankings, variables beyond the top 41 show a notable drop in predictive value, with phi-k correlations falling from 0.028 to 0.018. Based on this natural break in the correlation strength distribution, we selected the top 41 variables as our final modeling set. This threshold captures all features with meaningful predictive signal while maintaining model parsimony and computational efficiency.

The strongest predictors are the `EXT_SOURCE_1`, `EXT_SOURCE_2` and `EXT_SOURCE_3` variables, representing normalized scores from external credit bureaus. Despite containing substantial missing values, with `EXT_SOURCE_1` at 56% missingness and `EXT_SOURCE_3` at 20%, these variables demonstrate the highest correlations with default risk. The remaining features include borrower profiling variables such as `YEARS_EMPLOYED`, `YEARS_BIRTH`, `OCCUPATION_TYPE`, `ORGANIZATION_TYPE` and `NAME_INCOME_TYPE`, alongside engineered aggregations from the bureau and previous application tables such as `bur_cnt_active`, `bur_has_history` and `prev_cnt_consumer_approved`. Together, this feature set balances statistical predictive power with business interpretability for our modelling.

##

In [5]:
top_features = target_corr['feature'].head(41).tolist()
eda_features = eda(home_credit, top_features, categorical=True).reset_index(drop=True)
eda_features

Variable  Missing_Count  Missing_Percentage  \
0                     OWN_CAR_AGE         202929               65.99   
1                    EXT_SOURCE_1         173378               56.38   
2                    EXT_SOURCE_3          60965               19.83   
3                  YEARS_EMPLOYED          55374               18.01   
4                    EXT_SOURCE_2            660                0.21   
5                 AMT_GOODS_PRICE            278                0.09   
6                     AMT_ANNUITY             12                0.00   
7         YEARS_LAST_PHONE_CHANGE              1                0.00   
8                      AMT_CREDIT              0                0.00   
9              YEARS_REGISTRATION              0                0.00   
10                    YEARS_BIRTH              0                0.00   
11               YEARS_ID_PUBLISH              0                0.00   
12     REGION_POPULATION_RELATIVE              0                0.00   
13              ORGANIZATION_TYPE              0                0.00   
14          prev_cnt_cash_refused              0                0.00   
15        HOUR_APPR_PROCESS_START              0                0.00   
16                 bur_cnt_active              0                0.00   
17     prev_cnt_revolving_refused              0                0.00   
18     prev_cnt_consumer_approved              0                0.00   
19                OCCUPATION_TYPE              0                0.00   
20    prev_cnt_revolving_canceled              0                0.00   
21               NAME_INCOME_TYPE              0                0.00   
22             NAME_FAMILY_STATUS              0                0.00   
23              NAME_HOUSING_TYPE              0                0.00   
24            NAME_EDUCATION_TYPE              0                0.00   
25    REGION_RATING_CLIENT_W_CITY              0                0.00   
26           REGION_RATING_CLIENT              0                0.00   
27                    CODE_GENDER              0                0.00   
28         REG_CITY_NOT_WORK_CITY              0                0.00   
29                 FLAG_EMP_PHONE              0                0.00   
30         REG_CITY_NOT_LIVE_CITY              0                0.00   
31                FLAG_DOCUMENT_3              0                0.00   
32  AMT_REQ_CREDIT_BUREAU_flag_na              0                0.00   
33        LIVE_CITY_NOT_WORK_CITY              0                0.00   
34             NAME_CONTRACT_TYPE              0                0.00   
35                bur_has_history              0                0.00   
36                FLAG_DOCUMENT_6              0                0.00   
37                FLAG_WORK_PHONE              0                0.00   
38                     FLAG_PHONE              0                0.00   
39                   FLAG_OWN_CAR              0                0.00   
40               prev_has_history              0                0.00   

    Distinct_Values_Count  
0                      62  
1                  114584  
2                     814  
3                     464  
4                  119828  
5                    1002  
6                   13672  
7                     117  
8                    5603  
9                     555  
10                    483  
11                    177  
12                     81  
13                     58  
14                     38  
15                     24  
16                     23  
17                     23  
18                     22  
19                     19  
20                     14  
21                      8  
22                      6  
23                      6  
24                      5  
25                      3  
26                      3  
27                      3  
28                      2  
29                      2  
30                      2  
31                      2  
32                      2  
33                      2  
34                  

### <u> 3. Categorical Feature Engineering</u>

In [6]:
# Create final dataset with top features and target variable
home_credit_final = home_credit[top_features + ['TARGET']].copy()

# Assignment of 4 XNA in CODE_GENDER into class mode 'F'
gender_mapping = {'XNA': 'F'}
home_credit_final['CODE_GENDER'] = home_credit_final['CODE_GENDER'].replace(gender_mapping)


# Consolidation of rare categories in 'NAME_INCOME_TYPE (500 obs.) into 'Other' to address sample size instability
income_mapping = {
    'Unemployed': 'Other',
    'Maternity leave': 'Other',
    'Student': 'Other',
    'Businessman': 'Other'
}

home_credit_final['NAME_INCOME_TYPE'] = home_credit_final['NAME_INCOME_TYPE'].replace(income_mapping)


In [7]:
# Imputation based on 01_eda.ipynb analysis (2obs. to mode category 'Married')
family_status_mapping = {'Unknown' : 'Married'}
home_credit_final['NAME_FAMILY_STATUS'] = home_credit_final['NAME_FAMILY_STATUS'].replace(family_status_mapping)

While `OCCUPATION_TYPE` contains 19 distinct professions, each category holds at least 500 observations with varying default rates ranging from 4.8% (Accountants) to 17.2% (Low-skill Laborers). This meaningful risk gradient across occupation categories suggests valuable predictive signal, warranting preservation of the full granularity without consolidation. All 19 occupation types are for this reason maintained for model development.

In [ ]:
# Imputation single missing value in YEARS_LAST_PHONE_CHANGE with median
home_credit_final['YEARS_LAST_PHONE_CHANGE'] = home_credit_final['YEARS_LAST_PHONE_CHANGE'].fillna(home_credit_final['YEARS_LAST_PHONE_CHANGE'].median())

# Imputation of 12 missing values in AMT_ANNUITY with median
home_credit_final['AMT_ANNUITY'] = home_credit_final['AMT_ANNUITY'].fillna(home_credit_final['AMT_ANNUITY'].median())

In [9]:
# Consolidation of rare categories (<500 obs.) into 'Other' to address sample size instability
organization_mapping = {
    'Industry: type 13': 'Other',      # 67 obs
    'Trade: type 5': 'Other',          # 49 obs  
    'Trade: type 4': 'Other',          # 64 obs
    'Religion': 'Other',               # 85 obs
    'Industry: type 10': 'Other',      # 109 obs
    'Industry: type 6': 'Other',       # 112 obs
    'Transport: type 1': 'Other',      # 201 obs
    'Cleaning': 'Other',               # 260 obs
    'Legal Services': 'Other',         # 305 obs
    'Mobile': 'Other',                 # 317 obs
    'Trade: type 1': 'Other',          # 348 obs 
    'Industry: type 12': 'Other',      # 369 obs 
    'Culture': 'Other',                # 379 obs 
    'Realtor': 'Other',                # 396 obs 
    'Advertising': 'Other',            # 429 obs 
    'Industry: type 2': 'Other',       # 458 obs 
}

home_credit_final['ORGANIZATION_TYPE'] = home_credit_final['ORGANIZATION_TYPE'].replace(organization_mapping)

### <u> 4. Multicollinearity across features</u>

In [11]:
# Check correlation between EXT_SOURCE variables
ext_sources = home_credit_final[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']]
ext_corr = ext_sources.corr()
print(ext_corr)

              EXT_SOURCE_1  EXT_SOURCE_2  EXT_SOURCE_3
EXT_SOURCE_1      1.000000      0.213982      0.186846
EXT_SOURCE_2      0.213982      1.000000      0.109167
EXT_SOURCE_3      0.186846      0.109167      1.000000


In [12]:
# Check phik correlation between EXT_SOURCE variables
phik_ext_sources = phik_corr.loc[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3'], ['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']]
phik_ext_sources

EXT_SOURCE_1  EXT_SOURCE_2  EXT_SOURCE_3
EXT_SOURCE_1       1.00000      0.239200      0.203640
EXT_SOURCE_2       0.23920      1.000000      0.125038
EXT_SOURCE_3       0.20364      0.125038      1.000000

<u>Comment:</u>

Given the strength of the EXT_SOURCE variables as top predictors, a multicollinearity analysis was conducted to assess potential redundancy. Both Pearson and phi-k correlations reveal low to moderate associations between the three external sources (ranging from 0.12 to 0.24), indicating absence of multicollinearity concerns. Each variable captures distinct aspects of creditworthiness from different bureau sources, warranting retention of all three features without need for dimensionality reduction techniques to address information redundancy.

In [13]:
# Extract final phik correlation matrix for top 41 features
phik_final = phik_corr.loc[top_features, top_features]

# Build diagonal mask
mask = np.triu(np.ones_like(phik_final, dtype=bool))
high_correlation = phik_final.where(~mask).stack()
high_correlation_pairs = high_correlation[high_correlation > 0.7].sort_values(ascending=False)

print("Feature pairs with correlation > 0.7:")
print(high_correlation_pairs)

Feature pairs with correlation > 0.7:
FLAG_EMP_PHONE               ORGANIZATION_TYPE                1.000000
                             NAME_INCOME_TYPE                 1.000000
REGION_RATING_CLIENT         REGION_RATING_CLIENT_W_CITY      0.998765
bur_has_history              AMT_REQ_CREDIT_BUREAU_flag_na    0.998621
AMT_CREDIT                   AMT_GOODS_PRICE                  0.984828
LIVE_CITY_NOT_WORK_CITY      REG_CITY_NOT_WORK_CITY           0.962696
FLAG_EMP_PHONE               YEARS_BIRTH                      0.911242
FLAG_DOCUMENT_6              FLAG_EMP_PHONE                   0.806903
ORGANIZATION_TYPE            OCCUPATION_TYPE                  0.802636
NAME_INCOME_TYPE             ORGANIZATION_TYPE                0.775911
FLAG_DOCUMENT_6              NAME_INCOME_TYPE                 0.774435
YEARS_ID_PUBLISH             YEARS_BIRTH                      0.761784
FLAG_EMP_PHONE               OCCUPATION_TYPE                  0.759266
FLAG_DOCUMENT_6              ORGANIZATI

In [14]:
# Feature to drop based on high correlation analysis
features_to_drop = ['AMT_GOODS_PRICE', 'AMT_REQ_CREDIT_BUREAU_flag_na',
                     'FLAG_EMP_PHONE', 'REGION_RATING_CLIENT',
                     'LIVE_CITY_NOT_WORK_CITY']

# Drop from dataframe
home_credit_final = home_credit_final.drop(columns=features_to_drop)

# Drop from list
final_features = [f for f in top_features if f not in features_to_drop]

<u>Comment:</u>

The analysis of feature intercorrelations reveals substantial multicollinearity among several selected variables, leading to the removal of five features due to information redundancy.

- `AMT_GOODS_PRICE` exhibits near-perfect correlation (0.985) with `AMT_CREDIT`, as it represents a subset capturing only the goods value for consumer loans while `AMT_CREDIT` reflects the total credit amount across all loan types. Given this redundancy and the presence of missing values in `AMT_GOODS_PRICE` for non-consumer loans, this feature is dropped.

- `AMT_REQ_CREDIT_BUREAU_flag_na` shows extremely high correlation (0.999) with `bur_has_history`, as both capture whether the applicant has prior credit bureau records through different mechanisms. The former flags missing enquiry counts while the latter directly indicates bureau history presence, making `AMT_REQ_CREDIT_BUREAU_flag_na` redundant.

- `FLAG_EMP_PHONE` demonstrates perfect correlation with `ORGANIZATION_TYPE` because work phone provision patterns are largely determined by organization sector. Since `ORGANIZATION_TYPE` captures this employment stability signal while also providing granular employer sector information, the binary flag becomes redundant.

- `REGION_RATING_CLIENT` and `REGION_RATING_CLIENT_W_CITY` share a correlation of 0.999, as the former is a broader regional rating while the latter incorporates city-level information. The city-weighted version is then retained as it carries more granular geographic risk signal as well as holding higher phi-k correlation with the target response.

- Finally, `LIVE_CITY_NOT_WORK_CITY` and `REG_CITY_NOT_WORK_CITY` correlate at 0.963, both capturing residential and work location mismatches. The latter is retained given its stronger phi-k correlation with the target.

These decisions reduce the feature set from 41 to 36 variables while preserving all unique predictive information captured in the input space.

### <u> 5. Data Type</u>

In this final step, we ensure that all features are expressed using appropriate data types, namely int or float for numerical variables and binary variables, and category for categorical variables.

In [15]:
cat_cols = home_credit_final.select_dtypes(include="object").columns

for col in cat_cols:
    home_credit_final[col] = home_credit_final[col].astype("category")


In [16]:
home_credit_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 307511 entries, 149977 to 309985
Data columns (total 37 columns):
 #   Column                       Non-Null Count   Dtype   
---  ------                       --------------   -----   
 0   EXT_SOURCE_3                 246546 non-null  float64 
 1   EXT_SOURCE_1                 134133 non-null  float64 
 2   EXT_SOURCE_2                 306851 non-null  float64 
 3   YEARS_EMPLOYED               252137 non-null  float64 
 4   YEARS_BIRTH                  307511 non-null  float64 
 5   OCCUPATION_TYPE              307511 non-null  category
 6   ORGANIZATION_TYPE            307511 non-null  category
 7   NAME_INCOME_TYPE             307511 non-null  category
 8   REG_CITY_NOT_WORK_CITY       307511 non-null  int64   
 9   YEARS_LAST_PHONE_CHANGE      307511 non-null  float64 
 10  REG_CITY_NOT_LIVE_CITY       307511 non-null  int64   
 11  FLAG_DOCUMENT_3              307511 non-null  int64   
 12  bur_cnt_active               307511 non-null

### <u> 5. Final dataset</u>

In [17]:
eda_final = eda(home_credit_final, final_features, categorical=True).reset_index(drop=True)
eda_final

Variable  Missing_Count  Missing_Percentage  \
0                   OWN_CAR_AGE         202929               65.99   
1                  EXT_SOURCE_1         173378               56.38   
2                  EXT_SOURCE_3          60965               19.83   
3                YEARS_EMPLOYED          55374               18.01   
4                  EXT_SOURCE_2            660                0.21   
5                   AMT_ANNUITY              0                0.00   
6                    AMT_CREDIT              0                0.00   
7            YEARS_REGISTRATION              0                0.00   
8                   YEARS_BIRTH              0                0.00   
9              YEARS_ID_PUBLISH              0                0.00   
10      YEARS_LAST_PHONE_CHANGE              0                0.00   
11   REGION_POPULATION_RELATIVE              0                0.00   
12            ORGANIZATION_TYPE              0                0.00   
13        prev_cnt_cash_refused              0                0.00   
14      HOUR_APPR_PROCESS_START              0                0.00   
15               bur_cnt_active              0                0.00   
16   prev_cnt_revolving_refused              0                0.00   
17   prev_cnt_consumer_approved              0                0.00   
18              OCCUPATION_TYPE              0                0.00   
19  prev_cnt_revolving_canceled              0                0.00   
20            NAME_HOUSING_TYPE              0                0.00   
21             NAME_INCOME_TYPE              0                0.00   
22           NAME_FAMILY_STATUS              0                0.00   
23          NAME_EDUCATION_TYPE              0                0.00   
24  REGION_RATING_CLIENT_W_CITY              0                0.00   
25       REG_CITY_NOT_WORK_CITY              0                0.00   
26       REG_CITY_NOT_LIVE_CITY              0                0.00   
27              FLAG_DOCUMENT_3              0                0.00   
28           NAME_CONTRACT_TYPE              0                0.00   
29              bur_has_history              0                0.00   
30              FLAG_DOCUMENT_6              0                0.00   
31              FLAG_WORK_PHONE              0                0.00   
32                   FLAG_PHONE              0                0.00   
33                 FLAG_OWN_CAR              0                0.00   
34                  CODE_GENDER              0                0.00   
35             prev_has_history              0                0.00   

    Distinct_Values_Count  
0                      62  
1                  114584  
2                     814  
3                     464  
4                  119828  
5                   13672  
6                    5603  
7                     555  
8                     483  
9                     177  
10                    117  
11                     81  
12                     42  
13                     38  
14                     24  
15                     23  
16                     23  
17                     22  
18                     19  
19                     14  
20                      6  
21                      5  
22                      5  
23                      5  
24                      3  
25                      2  
26                      2  
27                      2  
28                      2  
29                      2  
30                      2  
31                      2  
32                      2  
33                      2  
34                      2  
35                      2

<u>Comment:</u>

While several features retain substantial missing values (`EXT_SOURCE_1` at 56%, `EXT_SOURCE_3` at 20%, `OWN_CAR_AGE` at 66%, and `YEARS_EMPLOYED` at 18%), these missingness patterns represent informative signals rather than data quality issues. For the external credit bureau scores, missing values indicate credit invisible applicants lacking prior bureau records. The `OWN_CAR_AGE` missingness flags applicants without vehicle ownership, while `YEARS_EMPLOYED` captures the sentinel value denoting non-employment status. Given the informative nature of these patterns and the native handling capabilities of tree-based algorithms, all missing values are preserved in their original state for model training. `AMT_ANNUITY	` is the only variable containing iniformantive missing value due to wrong or missing computation, but given the boosted tree model employed, this small fraction can be neglected

In [18]:
# Save final dataset for model deployment
home_credit_final.to_csv('../data/processed/home_credit_final.csv')